src/storage/table/variant_column_data.cpp

```diff
@@ -0,0 +1,454 @@
+		VariantColumnData::UnshredVariantData(intermediate, result, target_count);
+		return scan_count;
+	}
+	auto scan_count = validity.Scan(transaction, vector_index, state.child_states[0], result, target_count);
+	sub_columns[0]->Scan(transaction, vector_index, state.child_states[1], result, target_count);
+	return scan_count;
+}
+
+idx_t VariantColumnData::ScanCommitted(idx_t vector_index, ColumnScanState &state, Vector &result, bool allow_updates,
+                                       idx_t target_count) {
+	if (is_shredded) {
+		child_list_t<LogicalType> child_types;
+		child_types.emplace_back("unshredded", sub_columns[0]->type);
+		child_types.emplace_back("shredded", sub_columns[1]->type);
+		auto intermediate_type = LogicalType::STRUCT(child_types);
+		Vector intermediate(intermediate_type, target_count);
+
+		auto &child_vectors = StructVector::GetEntries(intermediate);
+		sub_columns[0]->ScanCommitted(vector_index, state.child_states[1], *child_vectors[0], allow_updates,
+		                              target_count);
+		auto scan_count = sub_columns[1]->ScanCommitted(vector_index, state.child_states[2], *child_vectors[1],
+		                                                allow_updates, target_count);
+
+		VariantColumnData::UnshredVariantData(intermediate, result, target_count);
+		return scan_count;
+	}
+	auto scan_count =
+	    sub_columns[0]->ScanCommitted(vector_index, state.child_states[1], result, allow_updates, target_count);
+	return scan_count;
+}
+
+idx_t VariantColumnData::ScanCount(ColumnScanState &state, Vector &result, idx_t count, idx_t result_offset) {
+	auto scan_count = sub_columns[0]->ScanCount(state.child_states[1], result, count, result_offset);
+	return scan_count;
+}
+
+void VariantColumnData::Skip(ColumnScanState &state, idx_t count) {
+	validity.Skip(state.child_states[0], count);
+
+	// skip inside the sub-columns
+	for (idx_t child_idx = 0; child_idx < sub_columns.size(); child_idx++) {
+		sub_columns[child_idx]->Skip(state.child_states[child_idx + 1], count);
+	}
+}
+
+void VariantColumnData::InitializeAppend(ColumnAppendState &state) {
+	ColumnAppendState validity_append;
+	validity.InitializeAppend(validity_append);
+	state.child_appends.push_back(std::move(validity_append));
+
+	D_ASSERT(!is_shredded && sub_columns.size() == 1);
+	for (idx_t i = 0; i < sub_columns.size(); i++) {
+		auto &sub_column = sub_columns[i];
+		ColumnAppendState child_append;
+		sub_column->InitializeAppend(child_append);
+		state.child_appends.push_back(std::move(child_append));
+	}
+}
+
+void VariantColumnData::Append(BaseStatistics &stats, ColumnAppendState &state, Vector &vector, idx_t count) {
+	if (vector.GetVectorType() != VectorType::FLAT_VECTOR) {
+		Vector append_vector(vector);
+		append_vector.Flatten(count);
+		Append(stats, state, append_vector, count);
+		return;
+	}
+
+	// append the null values
+	validity.Append(stats, state.child_appends[0], vector, count);
+
+	D_ASSERT(!is_shredded && sub_columns.size() == 1);
```

Comment:
We can't assume this, hitting this in verification runs right now
In `./build/reldebug/test/unittest --test-config test/configs/wal_verification.json test/sql/storage/types/variant/big_table.test`